featurebyte/routes/common/feature_or_target_table.py

```diff
@@ -0,0 +1,270 @@
+"""
+Feature or target table controller
+"""
+from __future__ import annotations
+
+from typing import Any, Generic, List, Optional, TypeVar
+
+from abc import abstractmethod
+from dataclasses import dataclass
+from http import HTTPStatus
+
+import pandas as pd
+from bson import ObjectId
+from fastapi import HTTPException, UploadFile
+
+from featurebyte.common.utils import dataframe_from_arrow_stream
+from featurebyte.models.base_feature_or_target_table import BaseFeatureOrTargetTableModel
+from featurebyte.models.feature_store import FeatureStoreModel
+from featurebyte.models.historical_feature_table import HistoricalFeatureTableModel
+from featurebyte.models.observation_table import ObservationTableModel
+from featurebyte.models.target_table import TargetTableModel
+from featurebyte.query_graph.graph import QueryGraph
+from featurebyte.query_graph.node import Node
+from featurebyte.routes.common.base import PaginatedDocument
+from featurebyte.routes.common.base_materialized_table import BaseMaterializedTableController
+from featurebyte.routes.task.controller import TaskController
+from featurebyte.schema.historical_feature_table import HistoricalFeatureTableCreate
+from featurebyte.schema.info import (
+    BaseFeatureOrTargetTableInfo,
+    HistoricalFeatureTableInfo,
+    TargetTableInfo,
+)
+from featurebyte.schema.target_table import TargetTableCreate
+from featurebyte.schema.task import Task
+from featurebyte.schema.worker.task.historical_feature_table import (
+    HistoricalFeatureTableTaskPayload,
+)
+from featurebyte.schema.worker.task.target_table import TargetTableTaskPayload
+from featurebyte.service.entity_validation import EntityValidationService
+from featurebyte.service.historical_feature_table import HistoricalFeatureTableService
+from featurebyte.service.observation_table import ObservationTableService
+from featurebyte.service.preview import PreviewService
+from featurebyte.service.target_table import TargetTableService
+
+InfoTypeT = TypeVar("InfoTypeT", HistoricalFeatureTableInfo, TargetTableInfo)
+
+MaterializedTableDocumentT = TypeVar(
+    "MaterializedTableDocumentT",
+    HistoricalFeatureTableModel,
+    TargetTableModel,
+)
+MaterializedTableDocumentServiceT = TypeVar(
+    "MaterializedTableDocumentServiceT",
+    HistoricalFeatureTableService,
+    TargetTableService,
+)
+PayloadT = TypeVar(
+    "PayloadT",
+    TargetTableTaskPayload,
+    HistoricalFeatureTableTaskPayload,
+)
+TableCreateT = TypeVar(
+    "TableCreateT",
+    TargetTableCreate,
+    HistoricalFeatureTableCreate,
```

Comment:
It seems a bit strange that all the derived classes have to be enumerated when defining the base class here.